File: JavaWrapper/src/jagger/Afster/Afl.java
package jagger.Afster;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.FileOutputStream;

public class Afl {
	static {
		loadLib("Afster");
        loadLib("AfsterNative");
    }

    private long _handle;
    private File _file;
    private int _entrycount;
    
    public final int AFL_NAMEBUFFERSIZE = 0x20;

    private static void loadLib(String libname) {
        if(System.getProperty("os.name").toLowerCase().contains("win")) {
            libname = "/" + libname + ".dll";
        } 
        else {
            libname = "/lib" + libname + ".so";
        }
    	try (InputStream in = Afl.class.getResourceAsStream(libname)) {
    		if(in == null) {
    			throw new UnsatisfiedLinkError(libname + " was not found.");
    		}
    		File tempLib = new File(System.getProperty("java.io.tmpdir"), libname.substring(1));
    		tempLib.deleteOnExit();
    		try (OutputStream out = new FileOutputStream(tempLib)) {
                byte[] buf = new byte[4096];
                int len;
                while ((len = in.read(buf)) > 0) {
                    out.write(buf, 0, len);
                }
            }
    		System.load(tempLib.getAbsolutePath());
    	}
    	catch (IOException ex) {
    		throw new RuntimeException(libname + " could not be loaded.");
    	}
    }
    
    private native long afl_open(String aflPath);
    private native long afl_create(long afs, String filepath);
    private native long afl_new(int entries, String filepath);
    private native String afl_getName(long afl, int id);
    private native int afl_getEntrycount(long afl);
    private native int afl_rename(long afl, int id, String new_name);
    private native void afl_free(long afl);
    private native int afl_save(long afl);
    private native int afl_saveNew(long afl, String filepath);
    private native int afl_importAfl(long afl, long afs, boolean permament);
    
    private native String[] jafl_getAllStrings(long afl);

    public Afl(String aflpath) {
        this._handle = afl_open(aflpath);
        this._file = new File(aflpath);
        this._entrycount = afl_getEntrycount(_handle);
    }
    
    public Afl(int amount_entries, String filepath) throws Exception {
    	if(amount_entries > 0xFFFF) {
    		throw new Exception("An AFL file can only have a maximum of 65535 entries.");
    	}
    	this._handle = afl_new(amount_entries, filepath);
    	this._file = new File(filepath);
    	this._entrycount = amount_entries;
    }
    
    public Afl(Afs afs, String filepath) {
    	this._handle = afl_create(afs._getHandle(), filepath);
    	this._file = new File(filepath);
    	this._entrycount = afl_getEntrycount(_handle);
    }

    public void free() {
        afl_free(_handle);
    }
    
    public long _getHandle() {
    	return _handle;
    }
    
    public String getName(int id) {
    	return afl_getName(_handle, id);
    }
    
    public int rename(int id, String new_name) {
    	return afl_rename(_handle, id, new_name);
    }
    
    public int getEntrycount() {
    	return _entrycount;
    }
    
    public int save() {
    	return afl_save(_handle);
    }
    
    public int saveNew(String filepath) {
    	return afl_saveNew(_handle, filepath);
    }
    
    public int importIntoAfs(Afs afs) {
    	return afl_importAfl(_handle, afs._getHandle(), afs.getPermanent());
    }
}

File: CSharpWrapper/Afl.cs
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace Afster
{
    internal static class AflNative
    {
        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate IntPtr DelegateAflOpen(string aflPath);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate IntPtr DelegateAflCreate(IntPtr afs, string filepath);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate IntPtr DelegateAflNew(ushort entries, string filepath);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate string DelegateAflGetName(IntPtr afl, int id);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int DelegateAflGetEntrycount(IntPtr afl);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate int DelegateAflRename(IntPtr afl, int id, string newName);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate void DelegateAflFree(IntPtr afl);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int DelegateAflSave(IntPtr afl);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
        public delegate int DelegateAflSaveNew(IntPtr afl, string filepath);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int DelegateAflImport(IntPtr afl, IntPtr afs, bool permament);

        // Function instances
        public static readonly DelegateAflOpen afl_open;
        public static readonly DelegateAflCreate afl_create;
        public static readonly DelegateAflNew afl_new;
        public static readonly DelegateAflGetName afl_getName;
        public static readonly DelegateAflGetEntrycount afl_getEntrycount;
        public static readonly DelegateAflRename afl_rename;
        public static readonly DelegateAflFree afl_free;
        public static readonly DelegateAflSave afl_save;
        public static readonly DelegateAflSaveNew afl_saveNew;
        public static readonly DelegateAflImport afl_importAfl;

        // Handle to the native library
        private static readonly IntPtr nativeHandle;

        // Helper to map function name to delegate
        private static T GetDelegate<T>(string name) where T : Delegate
        {
            IntPtr ptr = NativeLibrary.GetExport(nativeHandle, name);
            return Marshal.GetDelegateForFunctionPointer<T>(ptr);
        }

        static AflNative()
        {
            string libPath;
            string folder = AppContext.BaseDirectory;
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                libPath = "Afster.dll";
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
                libPath = "libAfster.so";
            else
                throw new PlatformNotSupportedException();

            libPath = Path.Combine(folder, libPath);   

            // Load the native library
            nativeHandle = NativeLibrary.Load(libPath);

            // Bind delegates to exported functions
            afl_open = GetDelegate<DelegateAflOpen>("afl_open");
            afl_create = GetDelegate<DelegateAflCreate>("afl_create");
            afl_new = GetDelegate<DelegateAflNew>("afl_new");
            afl_getName = GetDelegate<DelegateAflGetName>("afl_getName");
            afl_getEntrycount = GetDelegate<DelegateAflGetEntrycount>("afl_getEntrycount");
            afl_rename = GetDelegate<DelegateAflRename>("afl_rename");
            afl_free = GetDelegate<DelegateAflFree>("afl_free");
            afl_save = GetDelegate<DelegateAflSave>("afl_save");
            afl_saveNew = GetDelegate<DelegateAflSaveNew>("afl_saveNew");
            afl_importAfl = GetDelegate<DelegateAflImport>("afl_importAfl");
        }
    }

    public sealed class Afl : IDisposable
    {
        public const int AFL_NAMEBUFFERSIZE = AfsEntryMetadata.AFSMETA_NAMEBUFFERSIZE;
        internal IntPtr _handle;
        private bool _disposed;
        private bool _opened;

        public bool Permanent;

        public string Path { get; private set; }

        public int EntryCount { get; private set; }

        private Afl()
        {
            
        }

        public Afl(string filepath)
        {
            _handle = AflNative.afl_open(filepath);
            if (_handle == IntPtr.Zero)
            {
                throw new Exception($"Failed to open AFL: {filepath}");
            }
            EntryCount = AflNative.afl_getEntrycount(_handle);
            Path = filepath;
            _opened = true;
        }

        public static Afl Create(Afs afs, string path)
        {
            Afl afl = new Afl();
            afl._handle = AflNative.afl_create(afs._handle, path);
            if (afl._handle == IntPtr.Zero)
            {
                throw new Exception($"Failed to create AFL: {path}");
            }
            afl.EntryCount = AflNative.afl_getEntrycount(afl._handle);
            afl.Path = path;
            afl._opened = true;
            return afl;
        }

        public static Afl New(ushort entryCount, string path)
        {
            Afl afl = new Afl();
            afl._handle = AflNative.afl_new(entryCount, path);
            if (afl._handle == IntPtr.Zero)
            {
                throw new Exception($"Failed to create AFL: {path}");
            }
            afl.EntryCount = entryCount;
            afl.Path = path;
            afl._opened = true;
            return afl;
        }

        public string GetName(int index)
        {
            if (index < 0 || index >= EntryCount)
            {
                throw new Exception($"Index {index} outside of AFS range.");
            }
            return AflNative.afl_getName(_handle, index);
        }

        public int Rename(int index, string newName)
        {
            if (index < 0 || index >= EntryCount)
            {
                throw new Exception($"Index {index} outside of AFL range.");
            }
            return AflNative.afl_rename(_handle, index, newName);
        }

        public int Save()
        {
            return AflNative.afl_save(_handle);
        }

        public int SaveNew(string filepath)
        {
            return AflNative.afl_saveNew(_handle, filepath);
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                if (_handle != IntPtr.Zero)
                {
                    AflNative.afl_free(_handle);
                    _handle = IntPtr.Zero;
                }
                _disposed = true;
                GC.SuppressFinalize(this);
            }
        }
    }
}